### PR TITLE
mission:reset mission after finished

### PR DESCRIPTION
**Pull request by jinchengde**
vehicle could reset mission after mission finished, change the flight mode to others and back to mission mode, no need disarm, the mission will reset.

### CODE REVIEW

**Comment by dagar** [comment]
Great! This looks like a safe and simple way to do it.

File: src/modules/navigator/mission.cpp [hunk 1]
```diff
@@ -1684,6 +1684,13 @@ Mission::need_to_reset_mission(bool active)
 	} else if (_navigator->get_vstatus()->arming_state == vehicle_status_s::ARMING_STATE_ARMED && active) {
 		/* mission is running, need reset after disarm */
 		_need_mission_reset = true;
+
+	} else if (_navigator->get_mission_result()->finished
```
Comment (dagar) [inline src/modules/navigator/mission.cpp]: We have to be a little careful here because other parts of navigator set the mission result as well.

Comment (jinchengde) [reply src/modules/navigator/mission.cpp]: could you help to give some prompt which we need to protect, do we need to confirm vehicle in loiter mode before we reset the mission?

Comment (TSC21) [reply src/modules/navigator/mission.cpp]: @jinchengde we need to make sure that the next state after the mission ends does not crash the vehicle. Example: make sure the vehicle changes to HOLD or to LAND.

Comment (jinchengde) [reply src/modules/navigator/mission.cpp]: OK，I will test for that, and I wish to have another PR for more modify

Comment (jinchengde) [reply src/modules/navigator/mission.cpp]: @TSC21 I had test in SITL as below
1. automatic takeoff and mission
2. mission end, the vehicle start to circle around the last waypoint
3. change the mode to hold, the vehicle stay current altitude and circle around
4. change the mode to mission, the vehicle fly to the first waypoint
so there is no crash after we change the mode after the mission ends

**Comment by Antiheavy** [comment]
I realize this is old and closed, but after a recent conversation with @jinchengde and after thinking about this a lot, I think this proposed change makes a lot of sense.  @jinchengde I recommend you rebase and reopen this PR.  I can't comment intelligently on the code, but the functionality makes sense.

**Pull request closed without merging**